addons/crm/static/src/js/crm_forecast_service.js

```diff
@@ -0,0 +1,213 @@
+     *                         [this.end] must be set in order to use [greedy === true]
+     * @returns {Object} modified context
+     */
+    getContext(context, greedy=undefined) {
+        if (typeof greedy === 'undefined') {
+            greedy = this.greedy;
+        }
+        var output = {
+            ...context,
+            fill_temporal: 1,
+            fill_temporal_from: this.start.format("YYYY-MM-DD"),
+            fill_temporal_min: this.min_period
+        };
+        if (greedy) {
+            output.fill_temporal_to = moment(this.end).subtract(1, 'day').format("YYYY-MM-DD");
+        }
+        this.greedy = this._get("greedy");
+        return output;
+    }
+    /**
+     * Adds one "granularity" period to [this.end], to expand the current forecast period
+     */
+    expand() {
+        this.end.add(1, `${this.granularity}s`);
+    }
+}
+
+/**
+ * Forecast Service
+ *
+ * This service will be used to generate or recover Forecast Periods
+ * 
+ * A specific forecast period configuration will always refer to the same instance
+ * unless reset is true
+ * 
+ */
+ var ForecastService = AbstractService.extend({
+    /**
+     * @override
+     */
+    start: function () {
+        this._super.apply(this, arguments);
+        this._forecastPeriods = {};
+    },
+    /**
+     * Get a forecast period according to the [config]
+     * 
+     * @param {Object} config
+     * @param {string} config.modelName
+     * @param {string} config.field
+     * @param {string} config.granularity can either be : day, week, month, quarter, year
+     * @param {integer} config.min_groups optional (defaults to 4), minimal amount of desired groups
+     * @param {boolean} config.reset optional (defaults to false), whether the forecast period should be reinstancied
+     * @returns 
+     */
+    get: function (config) {
+        config = this._parseConfig(config);
+        if (!(config.modelName in this._forecastPeriods)) {
+            this._forecastPeriods[config.modelName] = {};
+        }
+        if (!(config.field in this._forecastPeriods[config.modelName])) {
+            this._forecastPeriods[config.modelName][config.field] = {};
+        }
+        if (!(config.granularity in this._forecastPeriods[config.modelName][config.field]) || config.reset) {
+            this._forecastPeriods[config.modelName][config.field][config.granularity] =
+                new ForecastPeriod(config.modelName, config.field, config.granularity, config.min_groups);
+        }
+        return this._forecastPeriods[config.modelName][config.field][config.granularity];
+    },
+    _parseConfig: function (config) {
+        return {
+            modelName: config.modelName,
+            field: config.field,
+            granularity: config.granularity,
+            min_groups: config.min_groups || 4,
+            reset: config.reset || false,
+        };
+    }
+});
```

Comment:
```suggestion
    _parseConfig({modelName, field, granularity, min_groups: 4, reset: false}) {
        return { modelName, field, granularity, min_groups, reset };
    }
});
```
mebbe? Unless e.g. a config with `{ min_groups: false }` should specifically be overridden?